From 3b39b2a5772ece85b866f4fa95256cdb19a18b68 Mon Sep 17 00:00:00 2001
From: Sureshkumar <suresure@qti.qualcomm.com>
Date: Thu, 22 Jan 2026 13:33:12 +0530
Subject: [PATCH] Fix unassociated STA link metrics issues

 This patch addresses two critical issues in unassociated STA link metrics:

 1. RCPI Double Conversion:
    - Fixed incorrect double conversion of RCPI values

 2. Unassociated STA Link Metrics Query Improvements:
    - Removed invalid empty UNASSOCIATED_STA_LINK_METRICS_QUERY_MESSAGE
      (not compliant with WiFi Alliance Multi-AP specification)
    - Added logic to skip sending queries to agents where the STA is
      already associated, avoiding redundant queries

Signed-off-by: Sureshkumar <suresure@qti.qualcomm.com>
---
 agent/src/beerocks/slave/son_slave_thread.cpp |  4 ++--
 .../master/tasks/link_metrics_task.cpp        | 20 ++++++++++++++++++-
 2 files changed, 21 insertions(+), 3 deletions(-)

diff --git a/agent/src/beerocks/slave/son_slave_thread.cpp b/agent/src/beerocks/slave/son_slave_thread.cpp
index 5f2c4dfd6..02abc64f0 100644
--- a/agent/src/beerocks/slave/son_slave_thread.cpp
+++ b/agent/src/beerocks/slave/son_slave_thread.cpp
@@ -3847,8 +3847,8 @@ bool slave_thread::handle_cmdu_monitor_message(const std::string &fronthaul_ifac
 
             stations_stats one_station_stats;
             one_station_stats.channel_number = station_in.channel;
-            one_station_stats.uplink_rcpi_dbm_enc =
-                wireless_utils::convert_rcpi_from_rssi(station_in.signal_strength);
+            // signal_strength is already in RCPI format
+            one_station_stats.uplink_rcpi_dbm_enc = station_in.signal_strength;
             one_station_stats.sta_mac = station_in.sta_mac;
             one_station_stats.measurement_to_report_delta_msec =
                 (uint32_t)std::chrono::duration_cast<std::chrono::milliseconds>(
diff --git a/controller/src/beerocks/master/tasks/link_metrics_task.cpp b/controller/src/beerocks/master/tasks/link_metrics_task.cpp
index 4e13486de..1444db590 100644
--- a/controller/src/beerocks/master/tasks/link_metrics_task.cpp
+++ b/controller/src/beerocks/master/tasks/link_metrics_task.cpp
@@ -63,7 +63,9 @@ void LinkMetricsTask::work()
             son_actions::send_cmdu_to_agent(agent->al_mac, cmdu_tx, database);
         }
 
-        // TODO: is the unassociated station request interval the same as the link metric ? or shall we consider a different one ?
+        // Periodic polling for UNASSOCIATED_STA_LINK_METRICS_QUERY
+        // is intended to be implemented as part of the strategy infrastructure
+        /*
         if (!cmdu_tx.create(
                 0, ieee1905_1::eMessageType::UNASSOCIATED_STA_LINK_METRICS_QUERY_MESSAGE)) {
             LOG(ERROR) << "Failed building message UNASSOCIATED_STA_LINK_METRICS_QUERY_MESSAGE!";
@@ -73,6 +75,7 @@ void LinkMetricsTask::work()
         for (const auto &agent : database.get_all_connected_agents()) {
             son_actions::send_cmdu_to_agent(agent->al_mac, cmdu_tx, database);
         }
+        */
 
         last_query_request = std::chrono::steady_clock::now();
     }
@@ -118,8 +121,23 @@ void LinkMetricsTask::handle_event(int event_enum_value, void *event_obj)
             return;
         }
 
+        // Retrieve station information to determine association status
+        auto station = database.get_station(unassoc_sta_link_metric_event->unassoc_sta_mac);
+
         // TODO: Check the capability of agent and send only to those which support unassoc sta link metric
         for (const auto &agent : database.get_all_connected_agents()) {
+            // Skip agents where the STA is already associated to avoid redundant queries
+            if (station && station->get_bss()) {
+                auto bss_parent_agent = database.get_bss_parent_agent(station->get_bss()->bssid);
+                if (bss_parent_agent == agent->al_mac) {
+                    LOG(DEBUG) << "Skipping agent " << agent->al_mac
+                               << " - STA " << unassoc_sta_link_metric_event->unassoc_sta_mac
+                               << " is already associated";
+                    continue;
+                }
+            }
+            LOG(DEBUG) << "Sending UNASSOCIATED_STA_LINK_METRICS_QUERY to agent " << agent->al_mac
+                       << " for STA " << unassoc_sta_link_metric_event->unassoc_sta_mac;
             son_actions::send_cmdu_to_agent(agent->al_mac, cmdu_tx, database);
         }
         break;
-- 
2.34.1

